qt-bugs@ issue : 
Trolltech task ID : 161944
bugs.kde.org number :
applied: yes
author: Andreas Aardal Hanssen <ahanssen@trolltech.com>


--- src/corelib/io/qprocess.cpp	Thu Jun 28 10:00:58 CEST 2007
+++ src/corelib/io/qprocess.cpp	Thu Jun 28 10:00:58 CEST 2007

@@ -816,15 +816,21 @@
     read(), readAll(), readLine(), and getChar(). It also determines
     which channel triggers QProcess to emit readyRead().
 
-    Changing the read channel will clear the unget buffer.
-
     \sa readChannel()
 */
 void QProcess::setReadChannel(ProcessChannel channel)
 {
     Q_D(QProcess);
-    if (d->processChannel != channel)
-        d->buffer.clear();
+    if (d->processChannel != channel) {
+        QByteArray buf = d->buffer.readAll();
+        if (d->processChannel == QProcess::StandardOutput) {
+            for (int i = buf.size() - 1; i >= 0; --i)
+                d->outputReadBuffer.ungetChar(buf.at(i));
+        } else {
+            for (int i = buf.size() - 1; i >= 0; --i)
+                d->errorReadBuffer.ungetChar(buf.at(i));
+        }
+    }
     d->processChannel = channel;
 }
 

